sympy/polys/matrices/domainmatrix.py

```diff
@@ -1692,6 +1750,582 @@ def det(self):
+
+        See Also
+        ========
+
+        eval_poly_mul
+        """
+        A = self
+        m, n = A.shape
+
+        if m != n:
+            raise DMNonSquareMatrixError("Matrix must be square")
+
+        if not p:
+            return self.zeros(self.shape, self.domain)
+        elif len(p) == 1:
+            return p[0] * self.eye(self.shape, self.domain)
+
+        # Evaluate p(A) using Horner's method:
+        # XXX: Use Paterson-Stockmeyer method?
+        I = A.I()
+        p_A = p[0] * I
+        for pi in p[1:]:
+            p_A = A*p_A + pi*I
+
+        return p_A
+
+    def eval_poly_mul(self, p, B):
+        r"""
+        Evaluate polynomial matrix product $p(A) \times B$.
+
+        Evaluate the polynomial matrix product $p(A) \times B$ using Horner's
+        method without creating the matrix $p(A)$ explicitly. If $B$ is a
+        column matrix then this method will only use matrix-vector multiplies
+        and no matrix-matrix multiplies are needed.
+
+        If $B$ is square or wide or if $A$ can be represented in a simpler
+        domain than $B$ then it might be faster to evaluate $p(A)$ explicitly
+        (see :func:`eval_poly`) and then multiply with $B$.
+
+        Examples
+        ========
+
+        >>> from sympy import QQ
+        >>> from sympy.polys.matrices import DM
+        >>> A = DM([[QQ(1), QQ(2)], [QQ(3), QQ(4)]], QQ)
+        >>> b = DM([[QQ(5)], [QQ(6)]], QQ)
+        >>> p = [QQ(1), QQ(2), QQ(3)]
+        >>> p_A_b = A.eval_poly_mul(p, b)
+        >>> p_A_b
+        DomainMatrix([[144], [303]], (2, 1), QQ)
+        >>> p_A_b == p[0]*A**2*b + p[1]*A*b + p[2]*b
+        True
+        >>> A.eval_poly_mul(p, b) == A.eval_poly(p)*b
+        True
+
+        See Also
+        ========
+
+        eval_poly
+        solve_den_charpoly
+        """
+        A = self
+        m, n = A.shape
+        mb, nb = B.shape
+
+        if m != n:
+            raise DMNonSquareMatrixError("Matrix must be square")
+
+        if mb != n:
+            raise DMShapeError("Matrices are not aligned")
+
+        if A.domain != B.domain:
+            raise DMDomainError("Matrices must have the same domain")
+
+        # Given a polynomial p(x) = p[0]*x^n + p[1]*x^(n-1) + ... + p[n-1]
+        # and matrices A and B we want to find
+        #
+        #   p(A)*B = p[0]*A^n*B + p[1]*A^(n-1)*B + ... + p[n-1]*B
+        #
+        # Factoring out A term by term we get
+        #
+        #   p(A)*B = A*(...A*(A*(A*(p[0]*B) + p[1]*B) + p[2]*B) + ...) + p[n-1]*B
+        #
+        # where each pair of brackets represents one iteration of the loop
+        # below starting from the innermost p[0]*B. If B is a column matrix
+        # then products like A*(...) are matrix-vector multiplies and products
+        # like p[i]*B are scalar-vector multiplies so there are no
+        # matrix-matrix multiplies.
+
+        p_A_B = p[0]*B
+
+        for p_i in p[1:]:
+            p_A_B = A*p_A_B + p_i*B
+
+        return p_A_B
```

Comment:
If the domain is EXRAW (i.e. Expr) then a factor might not be distributed. The other domains all distribute everything.

The implementation detail that we use the Horner method makes a big difference to performance. We avoid matrix-matrix multiplies because it is the Horner method applied to the product `p(A)*b`:
```python
In [529]: M = randMatrix(100).to_DM()

In [530]: b = randMatrix(100, 1).to_DM()

In [531]: cp = M.charpoly()

In [532]: %time ok = M.eval_poly_mul(cp, b)
CPU times: user 760 ms, sys: 0 ns, total: 760 ms
Wall time: 759 ms

In [534]: %time ok = M.eval_poly(cp) * b
CPU times: user 27 s, sys: 0 ns, total: 27 s
Wall time: 27 s
```
Note in the loop that if `B` was a column vector then `p_A_B` will always be a column vector. The produce `A*p_A_B` is an `O(n^2)` matrix-vector multiply and the product `p_i*B` is an `O(n)` scalar-vector multiply. A single matrix-matrix multiply is `O(n^3)` and here we would do `n` multiplies in a loop to evaluate `p(A)`. That means that `M.eval_poly(cp) * b` is `O(n^4)` arithmetic operations whereas `M.eval_poly_mul(cp, b)` is `O(n^3)` operations.